enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,188 @@
+workloads while they are transferred to other nodes.
+
+### Goals
+
+- Automatically pause all machineHealthChecks during cluster upgrades
+- Automatically unpause all machineHealthChecks when cluster upgrade finished
+- Get a simple solution done in time for OCP 4.9, by using what we have upstream already (the pause annotation), for
+  avoiding worst case behaviour during cluster upgrades. Look for better solutions for future releases
+
+### Non-Goals
+
+- Track the status of each single node (upgrading or not upgrading) for more fine-grained pausing of remediation
+
+## Proposal
+
+Introduce a new controller, which watches clusterVersion and machineHealthCheck resources, and pauses all
+machineHealthChecks when the clusterVersion status indicates an ongoing cluster upgrade. The controller will unpause the
+machineHealthChecks when the clusterVersion status indicates that the upgrade process finished.
+
+### User Stories
+
+- As a cluster admin I don't want unneeded reboots of nodes during cluster upgrades
+- As a cluster admin I don't want to manually pause machineHealthChecks during cluster upgrades
+- As a user I don't want my workloads to be transferred to new nodes more often than needed during cluster upgrades
+
+### Implementation Details/Notes/Constraints [optional]
+
+The clusterVersion object has several conditions in its status. Example:
+
+```yaml
+  status:
+    conditions:
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Done applying 4.9.0-0.nightly-2021-07-05-092650
+      status: "True"
+      type: Available
+    - lastTransitionTime: "2021-07-06T12:03:02Z"
+      status: "False"
+      type: Failing
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Cluster version is 4.9.0-0.nightly-2021-07-05-092650
+      status: "False"
+      type: Progressing
+    - lastTransitionTime: "2021-07-05T20:05:15Z"
+      message: 'Unable to retrieve available updates: currently reconciling cluster
+        version 4.9.0-0.nightly-2021-07-05-092650 not found in the "stable-4.8" channel'
+      reason: VersionNotFound
+      status: "False"
+      type: RetrievedUpdates
+    ...
+```
+
+The machineHealthCheck supports pausing by setting an annotation on it:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: openshift-machine-api
+  annotations:
+    cluster.x-k8s.io/paused: ""
+  ...
+```
+
+The new controller will add the pause annotation on the machineHealthChecks when the clusterVersion's `Progressing`
+condition switches to `"True"`. When that condition switches back to `"False"`, the controller will remove the
+annotation.
+
+The new controller will be implemented in the `openshift/machine-api-operator` repository, and be deployed in the
+`machine-api-controllers` pod, both alongside the machine-healthcheck-controller.
+
+### Risks and Mitigations
+
+Nothing we are aware of today
```

Comment:
> Hosts that are not successfully rebooted as part of the upgrade won't be remediated.

Depends on the status of the clusterVersion (or machineConfigPool as proposed at another place): will it continue to indicate an ongoing upgrade forever in case not all hosts rebooted successfully? 🤔 

Not sure if I understand the race... do you mean a cluster upgrade starts, and a host gets unhealthy at the same time for another reason? Or because of the cluster upgrade? The latter won't be an issue IMHO, because remediation only starts after some configured timeout, so there is enough time to pause the MHC.